Webapp upgrade protocol: Disable HTTP caching and reload other browser tabs to prevent fatal errors after new deployments.

adamziel opened the pull request:
## Motivation for the change, related issues

Solves fatal Playground breakages after new version deployments by adopting the following version upgrade protocol:

* Playground version is upgraded as early as possible after a new release
* HTTP cache is skipped when fetching new assets
* Stale Playground tabs are forcibly refreshed

Related to https://github.com/WordPress/wordpress-playground/issues/1821

## The problem

Playground got affected by HTTP caching and ended up loading assets from both the old release and the new release. This broke the app's dependency graph and led to fatal errors.

See the visualisation below. When Playground v184 is released, the app will only work properly if all the loaded assets come from v184:

![371781531-608a780e-60b8-4ed4-969a-d7497c7500a7](https://github.com/user-attachments/assets/605cba58-8eba-4fdb-b527-8c1f6942ce24)

## The solution

This PR ensures HTTP cache is skipped for assets that are cached offline. This isn't perfect as the browser will sometimes download the same file twice, but it's much better than breaking the app. We'll explore making the most out of both cache layers in the future.

Here's a rundown of the caching strategy implemented in this PR:

* Playground version is upgraded as early as possible after a new release
* HTTP cache is skipped

### Playground version is upgraded as early as possible after a new release

 New service workers call .skipWaiting(), immediately claim all the clients
 that were controlled by the previous service worker, and forcibly refreshes
 them.

 Why?

 Because Playground fetches new resources asynchronously and on demand. However,
 deploying a new webapp version of the app destroys the resources referenced in
 the previous webapp version. Therefore, we can't allow the previous version
 to run when a new version becomes available.

#### Push notifications

 It would be supremely useful to proactively notify the webapp after a fresh deployment.
 Playground doesn't do that yet but it likely will in the future.

### HTTP cache is skipped

 Playground relies on the **Cache only** strategy. It loads assets from
 the network, caches them, and serves them from the cache. The assumption
 is that all network requests yield the most recent version of the remote file.

 This helps us avoid the HTTP cache problem.

#### Cache layers

 We're dealing with the following cache layers:

 * HTTP cache in the browser
 * CacheStorage in the service worker
 * Edge Cache on playground.wordpress.net

#### HTTP cache in the browser

 This service worker skips the browser HTTP cache for all network requests. This is because
 the HTTP cache caused a particularly nasty problem in Playground deployments.

 Installing a new service worker purged the CacheStorage and requested a new set of assets
 from the network. However, some of these requests were served from the HTTP cache. As a
 result, Playground would start loading a mix of old and new assets and quickly error out.
 What made it worse is that this broken state was cached in CacheStorage, breaking Playground
 for weeks until the cache was refreshed.

 See https://github.com/WordPress/wordpress-playground/pull/1822 for more details.

#### CacheStorage in the service worker

 This servive worker uses a **Cache only** strategy to ensure all the loaded assets
 come from the same webapp build.

 The **Cache only** strategy means Playground only loads each assets from
 the network once, caches it, and serves it from the cache from that point on.

 The only times Playground reaches to the network are:

 * Before the service worker is installed.
 * When the service worker is being activated.
 * On CacheStorage cache miss occurs.

### Edge Cache on playground.wordpress.net

 The remote server (playground.wordpress.net) has an Edge Cache that's populated with
 all static assets on every webapp deployment. All the assets served by playground.wordpress.net
 at any point in time come from the same build and are consistent with each other. The
 deployment process is atomic-ish so the server should never expose a mix of old and new
 assets.

 However, what if a new webapp version is deployed right when someone downloaded 10 out of
 27 static assets required to boot Playground?

 Right now, they'd end up in an undefined state and likely see an error. Then, on a page refresh,
 they'd pick up a new service worker that would purge the stale assets and boot the new webapp
 version.

 This is not a big problem for now, but it's also not the best user experience. This can be
 eventually solved with push notifications. A new deployment would notify all the active
 clients to upgrade and pick up the new assets.

 ## Other changes

In addition, this PR:

* Adds E2E tests for app deployments and offline mode
* Updates CI to run Playwright tests:
	* Firefox
	* Safari
	* Chrome
* Fixes a few paper cuts
	* Fixed: Boot halted when OPFS isn't available due to error/success hooks never running (4e0ef74)
	* Fixed: "Save in this browser" option stays available even when there's no OPFS support (f6225a9)

## Paths not taken

* Relying on build-time hashes in the filenames for all caching. We can't rely on that for the most important routes: `/`, `/index.html`, `/remote.html`, `/sw.js` – they need stable URLs for multiple reasons.
* A different caching strategy, such as [network falling back to cache](https://web.dev/articles/offline-cookbook#network-falling-back-to-cache). 

## Caveats and follow-up work

* Let's find a way to leverage HTTP cache without breaking the offline cache.
* There's no way to recover from a deployment happening during a page load – let's fix it.
* A new service worker forcibly reloads other browser tabs and destroys their in-memory context. Let's solve it by storing temporary sites in OPFS.

## Testing Instructions (or ideally a Blueprint)

CI.

Yes, it sounds like a lame testing plan fur such a profound change. However, almost none of these changes can be tested in a local dev environment and a large part of this work was about covering app deployment in our E2E tests. 

If you want to try these tests locally and see what they do, you'll need this special setup:

```bash
$ npx nx e2e:playwright:prepare-app-deploy-and-offline-mode playground-website
$ npx playwright test --config=packages/playground/website/playwright/playwright.config.ts ./packages/playground/website/playwright/e2e/deployment.spec.ts --ui
```

## Related resources

 * PR that turned off HTTP caching: https://github.com/WordPress/wordpress-playground/pull/1822
 * Exploring all the cache layers: https://github.com/WordPress/wordpress-playground/issues/1774
 * Cache only strategy: https://web.dev/articles/offline-cookbook#cache-only
 * Service worker caching and HTTP caching: https://web.dev/articles/service-worker-caching-and-http-caching

bgrgicak commented:
I can confirm that test are failing in the E2E environment. When I try it manually locally in Chrome it works, if I disable _Update_ on refresh_ in the SW settings, it stops working.

I will explore different ways of busting cache tomorrow.

bgrgicak commented on packages/playground/website/playwright/deploy_test_setup.sh:
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
+
+# Store the current branch name
+CURRENT_BRANCH=$(git rev-parse --abbrev-ref HEAD)
+
+git checkout 4cbd5ac28
+PLAYGROUND_URL=http://localhost:8000 npx nx run playground-website:build:wasm-wordpress-net
+rm -rf dist/packages/playground/wasm-wordpress-net-old
+mv dist/packages/playground/wasm-wordpress-net dist/packages/playground/wasm-wordpress-net-old
+
+git checkout d7b1a7274
+PLAYGROUND_URL=http://localhost:8000 npx nx run playground-website:build:wasm-wordpress-net
+rm -rf dist/packages/playground/wasm-wordpress-net-new
+mv dist/packages/playground/wasm-wordpress-net dist/packages/playground/wasm-wordpress-net-new
+
+# Checkout back to the original branch
+git checkout $CURRENT_BRANCH
```
This script doesn't work for me, but I was able to use the code to generate local folders with it.

My suggestion is that we add an old version as a E2E fixture, similarly [to this Gutenberg file](https://github.com/WordPress/wordpress-playground/blob/reliable-deployments-sw/dist/packages/playground/website/test-fixtures/site-with-unpatched-gutenberg.zip).

I plan to take a look at this after the tests start passing.

brandonpayton commented on packages/playground/remote/service-worker.ts:
```diff
@@ -25,6 +25,53 @@ if (!(self as any).document) {
+	/**
+	 * Skip over the "waiting" lifecycle state, to ensure that our
+	 * new service worker is activated immediately when we only
+	 * have a single open tab.
```
Is this a stale comment? Where in this code do we only skip waiting when there is a single open tab?

adamziel replied:
yeah it's stale

brandonpayton replied:
> yeah it's stale

Cool. Updated!

brandonpayton commented on packages/playground/remote/src/lib/offline-mode-cache.ts:
```diff
@@ -23,21 +23,13 @@ export class OfflineModeCache {
 		}
 
 		let response = await this.cache.match(request, { ignoreSearch: true });
+		// @TODO: this.cache.match() returns stale data even after the cache has been purged
```
Is this something you've observed in testing?

adamziel replied:
Yeah but that was because the stale HTTP cache caused the old assets to be cached again under a fresh key. The match() calls seems fine.

brandonpayton replied:
> Yeah but that was because the stale HTTP cache caused the old assets to be cached again under a fresh key. The match() calls seems fine.

Ah, that makes sense. Good to hear. The TODO is now removed.

brandonpayton commented on packages/playground/remote/src/lib/offline-mode-cache.ts:
```diff
@@ -61,6 +60,27 @@ export class OfflineModeCache {
+			(key) =>
+				key.startsWith(CACHE_NAME_PREFIX) && key !== LATEST_CACHE_NAME
+		);
+		return Promise.all(oldKeys.map((key) => caches.delete(key)));
```
Is there a reason we aren't just using [CacheStorage.delete()](https://developer.mozilla.org/en-US/docs/Web/API/CacheStorage/delete)? Are there other things that may be cached that we want to respect?

adamziel replied:
Oh such a good idea!

brandonpayton replied:
>> Is there a reason we aren't just using CacheStorage.delete()? Are there other things that may be cached that we want to respect?
>
> Oh such a good idea!

😅 We already were using CacheStorage.delete() here, but I had misread. It was such a good idea that we were already doing it.

brandonpayton commented on packages/playground/website/playwright/e2e/deployment.spec.ts:
```diff
@@ -0,0 +1,153 @@
+};
+
+const startServer = async (page, path: string): Promise<Server> => {
+	const cp = spawn('python3', ['-m', 'http.server', port.toString()], {
```
What if we used `php -S` here instead as this project has more of a relationship with PHP than Python. Maybe fewer would need to install a Python dependency.

Alternately, is there a reason we shouldn't just run a node-based HTTP server like [`serve`](https://www.npmjs.com/package/serve)?

brandonpayton replied:
I was having trouble getting the tests to run because of ENOENT errors when `python3` couldn't be found. It must have been some kind of pathing issue because `python3` is in my PATH. The same thing happened for `php`.

So I went ahead and switched to an inline Node.js-based HTTP server with https://github.com/WordPress/wordpress-playground/pull/1822/commits/fadf6e97b6cbdb868006ba3410e68766726c3f2f. It seems like this will be more reliable for more folks than using Python or PHP installed separately. Please feel free to revert if needed though.

brandonpayton commented:
I am able to reproduce the Playwright failures locally. After loading from the "new" files, the Playground iframe looks blank:
![Screenshot 2024-10-01 at 12 10 37 AM](https://github.com/user-attachments/assets/c6023a56-0c0e-450c-a771-d7b14597b82b)

I'm not sure why yet.

One thing I wonder is how much this could be interacting with the regular browser cache. If that is a possible confounding factor, what if we use the Service Worker to add cache-disabling response headers to our responses to the browser? Could that effectively eliminate that layer so that the Service Worker would have full control of all cached app resources?

brandonpayton commented:
As an aside, Playwright's UI mode is handy. I like it!
```
npx nx run playground-website:e2e:playwright -- --ui
```

It gives you full control of which tests to run and a nice, browsable display of results and screen captures along the way.

bgrgicak commented:
> As an aside, Playwright's UI mode is handy. I like it!

You can also add `--debug` for extra errors.

adamziel commented:
@brandonpayton HTTP caching is spot on! I updated the tests to:

* Run once with HTTP caching disabled – this passes
* Run once with HTTP caching enabled – this fails

I also added a `/switch-to-new-version` endpoint to avoid messing with the server process in the middle of the test.

Next, I logged the output from the python server for both runs: https://gist.github.com/adamziel/ff50b5fb945840c64b06b83ea373a936.

With caching disabled, the browser requests the same file a few times over. No surprises there.

However, with caching enabled, the browser only re-requests the following files:

```
127.0.0.1 - - 🕙 "GET /switch-to-new-version HTTP/1.1" 200 -
127.0.0.1 - - 🕙 "GET /sw.js HTTP/1.1" 200 -
127.0.0.1 - - 🕙 "GET /wp-6.6/wordpress-static.zip HTTP/1.1" 200 -
127.0.0.1 - - 🕙 "GET / HTTP/1.1" 200 -
127.0.0.1 - - 🕙 "GET /assets/index-191909dc.js HTTP/1.1" 200 -
127.0.0.1 - - 🕙 "GET /assets/index-338819e8.css HTTP/1.1" 200 -
127.0.0.1 - - 🕙 "GET /assets/main-25d144ec.css HTTP/1.1" 200 -
127.0.0.1 - - 🕙 "GET /assets/main-cf01976e.js HTTP/1.1" 200 -
127.0.0.1 - - 🕙 "GET /assets/client-22d3bbd1.js HTTP/1.1" 200 -
127.0.0.1 - - 🕙 "GET /assets/config-3ed3fcde.js HTTP/1.1" 200 -
127.0.0.1 - - 🕙 "GET /sw.js HTTP/1.1" 304 -
127.0.0.1 - - 🕙 code 404, message File not found
127.0.0.1 - - 🕙 "GET /assets/wp-6.6-917ecf17.zip HTTP/1.1" 404 -
127.0.0.1 - - 🕙 code 404, message File not found
127.0.0.1 - - 🕙 "GET /assets/php_8_3-4d8e97bf.js HTTP/1.1" 404 -
127.0.0.1 - - 🕙 "GET /sw.js HTTP/1.1" 304 -
```

It doesn't retrieve the updated `/index.html`, `remote.html`, or `/assets-required-for-offline-mode.json`.

adamziel commented:
I also found that replacing `cachedFetch()` (that caches the responses in a named Cache instance) with just `fetch()` solves the problem as well – even when HTTP caching is enabled.

adamziel commented:
I started logging Cache name every time a response is served from offline Cache.

* When HTTP caching is enabled, the same cache is used before and after the new app version is deployed.
* When HTTP caching is disabled, only the new cache ID is used after the new app version is deployed.

That would happen without refreshing the service worker, but the problem is solved if I only replace `cachedFetch()` with `fetch()` in the new service worker and leave the old one unchanged.

adamziel commented:
Relevant article: https://web.dev/articles/service-worker-caching-and-http-caching

adamziel commented:
The test passes when the server sends no-cache headers for `/remote.html` and `/index.html`, even if all the other files are HTTP-cached. Why?

brandonpayton commented:
> The test passes when the server sends no-cache headers for `/remote.html` and `/index.html`, even if all the other files are HTTP-cached. Why?

@adamziel, maybe because both use build-specific script resources:
- `index.html` asks for `<script type="module" crossorigin src="/assets/index-f9b131ee.js"></script>`
- `remote.html` asks for `<script type="module" crossorigin src="/assets/wordpress-5b0e2fb1.js"></script>`

brandonpayton commented:
And once the browser sees those pages updated and loads those scripts, it finds that those scripts reference dependencies it has never cached, and it requests them from the web server.

brandonpayton commented:
> @brandonpayton HTTP caching is spot on! I updated the tests to:
> 
> - Run once with HTTP caching disabled – this passes
> - Run once with HTTP caching enabled – this fails

This is great to know!

adamziel commented:
Here's what I figured out so far:

1. The browser requests the old Playground version
2. Assets are downloaded
3. The service worker is installed (hash=4cbd5ac28cf1d7d9588a1e90f5349815a5562d2a)
4. The service worker requests all the "offline mode" assets to store in the offline Cache (key=playground-cache-4cbd5ac28cf1d7d9588a1e90f5349815a5562d2a)
5. We switch to the new app version and refresh the page
6. The old index.html is loaded with all its related assets served from the old cache
7. The browser notices a new service worker and runs the installation hook
8. The new worker calls `.skipWaiting()`
9. The new worker activation event is dispatched
10. The old cache is purged
11. The new service worker starts populating the new cache with assets (key=playground-cache-66b55041575d5f12ec5ad41f4006ae4f6d3f46bc)
    * **Any `fetch()` that reuses a previously cached HTTP response in this step will break the dependency graph. We'd end up with some assets from the new Playground version and some assets from the old Playground version.**
13. The new service worker claims and refreshes every Playground browser tab
14. The new Playground version is loaded

There may be more.

adamziel commented:
Adding this header to every `fetch()` initiated by the service worker seems to help:

```
			headers: {
				"Cache-Control": "no-cache"
			}
```

Here's what that looks like in the minified version:

```js
const uncachedFetch = (url, options = {}) => {
	// Logging 😅 
	fetch('/?C=' + C);
	return fetch(url, {
		...options,
		cache: "no-store",
		headers: {
			"Cache-Control": "no-cache"
		}
	})
}
class y {
    constructor(e) {
        this.hostname = self.location.hostname, this.cache = e
    }
    static async getInstance() {
        if (!y.instance) {
            const e = await caches.open(C);
            y.instance = new y(e)
        }
        return y.instance
    }
    async cachedFetch(e) {
        if (!this.shouldCacheUrl(new URL(e.url))) return await uncachedFetch(e);
        let t = await this.cache.match(e, {
            ignoreSearch: !0
        });
		if (t) {
			fetch(`/cache-hit-NEW-from-${C}-for-${e.url}`);
		}
        return t || (t = await uncachedFetch(e), t.ok && self.serviceWorker.state === "activated" && await this.cache.put(e, t.clone())), t
    }
    async cacheOfflineModeAssetsForCurrentRelease() {
        if (!this.shouldCacheUrl(new URL(location.href))) return;
		const t = await (await uncachedFetch("/assets-required-for-offline-mode.json")).json();
		const urls = [...t, "/"].map(url => new Request(url, {
			method: "GET",
			headers: {
				"Cache-Control": "no-cache"
			}
		}));
        await this.cache.addAll(urls)
    }
```

adamziel commented:
Oh, if the server provides caching headers for `sw.js` then it breaks, too. I'll look for the right place in the client app to turn off caching for that URL.

adamziel commented:
This is much more reliable in skipping the cache than setting the `Cache-Control` header:

```js
	return fetch(url, {
		...options,
		cache: "no-cache",
	})
```

brandonpayton commented:
> This is much more reliable in skipping the cache than setting the `Cache-Control` header:
> 
> ```js
> 	return fetch(url, {
> 		...options,
> 		cache: "no-cache",
> 	})
> ```

Interesting! Out of curiosity, does the `cache: "no-store" give us the same result? I can test...

brandonpayton commented:
My earlier difficulty getting `python3` to run seems to be involved with using `npx`. Once I started invoking `nx` directly in my shell, it started taking advantage of my PATH and found `python3` with no trouble.

I think maybe we should adjust this to spawn a Node.js based server so we can just install everything needed through a JS dev dependency, but the python server is working just fine for now.

brandonpayton commented:
I've been having trouble getting the build portion of this to complete without errors. It keeps complaining about "permission denied" when trying to remove previous versions of old/new build. Maybe the latest changes with the zipped old build will fix this. Otherwise, I'm looking into it.

brandonpayton commented:
One thing I'm wondering:
Do we just need uncached fetch when requesting assets named in PWA manifest?

Is it a mistake to skip the browser cache completely like I think we are doing in https://github.com/WordPress/wordpress-playground/pull/1822/commits/7411b817e3b76fb9d0d9cd067d2056a1879b00f9? I would guess it is a mistake but haven't thought through it.

On the other hand, maybe the app should take complete control of how all Playground-requested resources are cached. Then we might have a better chance at understanding which things are coming from cache and why.

brandonpayton commented:
@adamziel you mentioned possibly moving OfflineModeCache things into the Service Worker module, and I took a look at that. I extracted the offline cache methods into top-level functions that could be used in the Service Worker, but at the last minute, I discovered that worker-utils.ts also has a dependency on OfflineModeCache.

So, I ended up keeping `offline-mode-cache.ts` and switching it to export separate, top-level functions because it seemed clearer to me. I also switched worker-utils to ask offline-mode-cache whether a URL is cached instead of opening and asking the cache directly. This way, all offline cache interactions are kept in the dedicated module. The change is https://github.com/WordPress/wordpress-playground/pull/1822/commits/6e29f3c49ce81a6b008f79be6eee0713bbdaf7a0. Please feel free to revert if you prefer. cc @bgrgicak

brandonpayton commented:
> Is it a mistake to skip the browser cache completely like I think we are doing in [7411b81](https://github.com/WordPress/wordpress-playground/commit/7411b817e3b76fb9d0d9cd067d2056a1879b00f9)? I would guess it is a mistake but haven't thought through it.

After more time reviewing the code, I think it probably makes sense to err on the side of fetching fresh resources for Playground, and if we need to back off from that a bit, we can.

brandonpayton commented:
I made a change to also delete individual entries from stale offline caches because, in testing, a reference to a deleted cache continued providing matched responses. https://github.com/WordPress/wordpress-playground/pull/1822/commits/9f06896c10693512889ad9070e414cfe0ada2af2

But on further consideration, that is probably a feature while previous versions of code are still running, and I don't know how it could do any harm. So I reverted that change.

brandonpayton commented on packages/playground/remote/service-worker.ts:
```diff
@@ -25,6 +29,57 @@ if (!(self as any).document) {
+					() => {
+						// @TODO: Only if we're running a version of Playground that
+						//        stores temporary sites in OPFS. Otherwise we'd
+						//        destroy the user's in-memory changes.
```
Even if all sites are being saved to OPFS, users could still have unsaved, in-memory changes while editing posts or something similar. Because of this, we might just need to ask users via prompts in Playground instances (#566).

adamziel replied:
It's not much of a choice, though. If the user says "no" in the prompt, their running app will be broken as they will no longer be able to, say, start a new site with another version of PHP as the relevant .wasm will no longer be available on the server. I'd rather make this "just work" and find a way to force refresh without losing data. One way of doing that would be storing all the "temporary" sites in OPFS anyway, only showing the ones created in the current browsing session, and cleaning the rest after 24h.

brandonpayton replied:
The thing is that we cannot force refresh without losing some data. 

If we store temporary sites in OPFS, we can preserve data that has been written to the file system, but forcing a refresh will still cause users to lose things they are working on that have not been saved.

Maybe that is a reasonable compromise. But I wanted to point that out.

It seems like we could inform users and offer them a choice about whether to delay updating temporarily. If they choose to delay, we could add a banner that an update is imminent and even disable things in the user interface that would require update. This could be a follow-up PR.

What do you think?

bgrgicak replied:
The more I think about this, the more I feel that the installation and update process needs to be fully controlled. 
If you would always get the cached version first we would ensure nothing breaks on the first load. 
After that, we can check for an update async and ask the user what to do.

I would do it automatically if it wasn't for their data, but we can't just remove someone's work just because it's not saved. (Imagine using Playground as your primary note taking app and it randomly reloads)

adamziel replied:
What kind of data loss are we risking? Posts have autosaves and revisions.

bgrgicak commented on packages/php-wasm/web-service-worker/src/initialize-service-worker.ts:
```diff
@@ -64,7 +64,8 @@ async function defaultRequestHandler(event: FetchEvent) {
 		});
-		return fetch(request);
+		// TODO: Is this necessary? Do we really want to bypass browser cache for all requests?
+		return fetchFresh(request);
```
It depends, this is a technically simple solution that will work, but it removes all the benefits of browser caching.

My problem with this is that the client takes over control of cache busting and we don't control a running client. 

I would prefer if we kept cache headers and focused on ensuring the version hash generated by builds works for all critical resources. 

Once this is working we can fetch by default. If the version is the same, the browser will return a cached resource.

If the connection is unstable, or offline, we let the service worker return the resource from cache, but otherwise, we don't rely on the service worker's cache.

adamziel replied:
> I would prefer if we kept cache headers and focused on ensuring the version hash generated by builds works for all critical resources.

We can't have a version hash for at least a few routes: `/sw.js`, `/index.html`, `/remote.html`, `/`, and probably some more like perhaps `/manifest.json`. I'm happy to elaborate if this seems weird.

bgrgicak commented:
> @adamziel, maybe because both use build-specific script resources:

This is connected to service worker cache. Whenever Remote is rebuilt we get a new `buildVersion`, this is used by [offline mode caching](https://github.com/WordPress/wordpress-playground/blob/d7b1a72745117ff24b55b048e061afefa2849673/packages/playground/remote/src/lib/offline-mode-cache.ts#L6) to create a cache name.

If the `buildVersion` is updated this will automatically bust all the cache.

bgrgicak commented:
Let me try and sum up what I learned so far.

> The [article @adamziel  shared](https://web.dev/articles/service-worker-caching-and-http-caching) describes different scenarios. 

By using the offline mode cache we are using the Long-term caching scenario which isn't a good fit for a frequently changing project like Playground and we should default to a version of Short-term caching.

> Current caching flow

- a request goes to offline mode cache in the service worker
  - if cache for the current `buildVersion` of the Remote package exists we attempt to fetch from cache
  - if the request was cached in the service worker it's returned from cache
  - if not it's fetched by the browser 
    - the browser decides based on cache headers if it will do a network request

This shows a couple of flaws.

First, we default to service worker cache while we should default to browser cache to keep caching periods short.

Second, we rely on the `buildVersion` to bust the service worker cache, but there is no clear control over it. 
We let Vite decide when to update it.

Even if Vite updates it, we don't explicitly fetch it every time, but allow the browser to decide based on cache headers.

adamziel commented:
> First, we default to service worker cache while we should default to browser cache to keep caching periods short.

How would it work in the offline mode work without defaulting to the service worker cache? Or in an online mode but on a shaky network? 

> Second, we rely on the buildVersion to bust the service worker cache, but there is no clear control over it. We let Vite decide when to update it.

`buildVersion` comes from a custom vite plugin that we control and reflects either the git ref of the current commit or the current date. It should be different for each build:

https://github.com/WordPress/wordpress-playground/blob/e37749388462c2877da90061774693a7a21329ca/packages/vite-extensions/vite-build-version.ts#L4-L18

> Even if Vite updates it, we don't explicitly fetch it every time, but allow the browser to decide based on cache headers.

What do you mean?

adamziel commented:
Good point about the http cache overall though. Hm, maybe we could have a special naming scheme for resources with content hash, such as index-h[bg4def]h.js and keep the http cache active for matching urls only?

bgrgicak commented:
> How would it work in the offline mode work without defaulting to the service worker cache? Or in an online mode but on a shaky network?

For offline it's simple `try { networkFetch } catch { cacheFetch }`.

Shaky networks are tricky, because we shouldn't mix `networkFetch` and `cacheFetch` otherwise Playground might break.

What if?
- we create a timestamp on every website deploy and use it as the version
- we fetch it as early as we can on boot and use that for cache control in the service worker
  - all files that are fetched before the version number should use the network first approach to caching in the service worker and have cache headers disabled
  - all files that are fetched after that use the cache first approach in the service worker and have cache headers enabled

What's missing here?

> buildVersion comes from a custom vite plugin that we control and reflects either the git ref of the current commit or the current date.

Good point. In that case we should find a solution for these files ([source](https://playground.wordpress.net/assets-required-for-offline-mode.json)). Should they have some form of cache busting? If yes can it be a dynamic name or version query string, if not should we disable cache headers.
```
  "/builder/builder.html",
  "/favicon.ico",
  "/gutenberg.css",
  "/gutenberg.html",
  "/index.html",
  "/logo-192.png",
  "/logo-256.png",
  "/logo-384.png",
  "/logo-512.png",
  "/manifest.json",
  "/ogimage.png",
  "/previewer.css",
  "/site-manager-background.svg",
  "/wordpress-importer.zip",
  "/wordpress.html",
  "/wordpress.svg",
  "/remote.html",
  "/sw.js",
```

> Even if Vite updates it, we don't explicitly fetch it every time, but allow the browser to decide based on cache headers.

This was related to t[he comment about why disabling cache headers fixes caching](https://github.com/WordPress/wordpress-playground/pull/1822).

From what I understand, if `remote.html` for example is fetched every time by disabling cache headers.
It will get the latest `buildVersion` and update service worker cache if needed.

This matches [what you found out ](https://github.com/WordPress/wordpress-playground/pull/1822#issuecomment-2385600396). _The test passes when the server sends no-cache headers for /remote.html and /index.html_

bgrgicak commented:
> Hm, maybe we could have a special naming scheme for resources with content hash, such as index-h[bg4def]h.js and keep the http cache active for matching urls only?

How would that be different from what we do with [assets-required-for-offline-mode.json](https://playground.wordpress.net/assets-required-for-offline-mode.json)?

adamziel commented:
> > Hm, maybe we could have a special naming scheme for resources with content hash, such as index-h[bg4def]h.js and keep the http cache active for matching urls only?
> 
> 
> 
> How would that be different from what we do with [assets-required-for-offline-mode.json](https://playground.wordpress.net/assets-required-for-offline-mode.json)?
> 
> 

1. It would also apply to files not explicitly listed in that json.
2. It would apply to files from the JSON list that are fetched before the JSON list is downloaded.

I'll reply to the other comment soon

adamziel commented:
Also, I'm tempted to get this in and move further http cache improvements to another pr. We'd make the webapp a bit slower to reload, but I don't see other downsides. Thoughts?

adamziel commented:
> For offline it's simple try { networkFetch } catch { cacheFetch }.
> Shaky networks are tricky, because we shouldn't mix networkFetch and cacheFetch otherwise Playground might break.

Hm. Let's break it down and see what comes out. We have the following ways of triggering network requests in Playground:

* **Direct document requests**, e.g. I type in `playground.wordpress.net` or `playground.wordpress.net/wordpress.html` in my address bar and press enter.
* **HTML** tags, e.g.`<img>`, `<link>`, `<script>`, `<iframe>` etc.
* **CSS** properties, e.g.`background-image` and webfonts
* **JavaScript** calls, both implicit and explicit, e.g. `fetch()`, `new Worker()`, `iframe.src = '';`, `serviceWorker.register()`, `cache.add()`, and likely many other ones

All of these can happen in the following contexts:

* Vanilla – First Playground page load, before the service worker is in place
* With service worker in place – after it's installed
* Within `<iframe>` on another domain
* ...probably a few more...

I'm not sure:

* How to reliably cover all of these with `try { networkFetch } catch { cacheFetch }`
* What to do when the 7th `import` fails assuming we've loaded the previous 6 from the network
* What would happen if the offline status changes while the app is open

My gut says we need access to a consistent dependency graph throughout the entire time the Playground app is open. The version on the server may change while we're working with Playground, and when that happens we can only do two things:

1. Keep sourcing what we have from the offline cache – and we probably won't have everything, e.g. php_7_2.wasm
2. Reload the page and fetch the new app version

adamziel commented:
[The offline cookbook](https://web.dev/articles/offline-cookbook) discusses a lot of strategies. You've described [network falling back to cache](https://web.dev/articles/offline-cookbook#network-falling-back-to-cache). Here's an excerpt from the description:

> This means you give **online users the most up-to-date content**, but offline users get an older cached version. If the network request succeeds you'll most likely want to [update the cache entry](https://web.dev/articles/offline-cookbook#on-network-response).

Most up-to-date content is a problem. Playground can only work if we ship _all content from the new version_ or _all content from the old version_, but not _fresh content from the new version, old content from cache_.

> However, this method has flaws. If the user has an intermittent or slow connection they'll have to wait for the network to fail before they get the perfectly acceptable content already on their device. This can take an extremely long time and is a frustrating user experience. See the next pattern, [Cache then network](https://web.dev/articles/offline-cookbook#cache-then-network), for a better solution.

This is a large downside.


I don't see a reliable way of doing any other strategy than [Cache only](https://web.dev/articles/offline-cookbook#cache-only):

> Ideal for: anything you'd consider static to a particular "version" of your site. You should have cached these in the install event, so you can depend on them being there.

I'm happy to be wrong here. Do you think we could do it in another way?

adamziel commented:
> Good point. In that case we should find a solution for these files ([source](https://playground.wordpress.net/assets-required-for-offline-mode.json)). Should they have some form of cache busting? If yes can it be a dynamic name or version query string, if not should we disable cache headers.

At least a few items on that list cannot have a dynamic name. AFAIR the service worker script must have a stable name. All `html` files need a stable URL for people linking to them and embedding them. I think manifest.json also needs to be stable, but I'm not 100% sure about that one. We need a caching strategy that works end to end without relying on the names. To me, the hashes in assets names are helpful as a debugging aid but not as a reliable caching strategy.

bgrgicak commented:
> Also, I'm tempted to get this in and move further http cache improvements to another pr. We'd make the webapp a bit slower to reload, but I don't see other downsides. Thoughts?

Let's do that. This PR is blocking all other development, and from the discussion, it doesn't seem like we will be able to resolve everything quickly.

bgrgicak commented:
> Ideal for: anything you'd consider static to a particular "version" of your site. You should have cached these in the install event, so you can depend on them being there.

That's us. In that case we fundamentals right in `trunk`. All requests go through cache until it's busted.

Now how to handle busting?

The most reliable way I can think of is to use the cached version, check if there is a new one, if yes offer the user to reload and update.

The only other option I can think of is to first load the version, check if it's different, bust cache, and reload if it is. 
This sounds more seamless, but it also creates a bottleneck, every page load would wait for the version to be fetched before doing anything else. 

What would be some other strategies?

adamziel commented:
> The most reliable way I can think of is to use the cached version, check if there is a new one, if yes offer the user to reload and update.

Yeah that's our only option for now. Technically, this PR gets us there. The new service worker handles cache invalidation and upgrading active clients to the new version as follows:

1. Load the old app version
2. Fetch the service worker
3. New service worker found, run the activation hooks
4. The activation hook purge the cache and reload every browser tab

Ideally we'd have something on the server to tell an active client "hey, there's a new version, you need to upgrade". Something like a websocket or a push notification. TBD.

Aha, page reloading is not much of a choice, see [this comment](https://github.com/WordPress/wordpress-playground/pull/1822#discussion_r1784012274). The best we can do is force-reload and ensure no in-memory state got destroyed in the process.

adamziel commented:
I think this is good to go once the e2e tests pass. Or even if they don't pass, assuming the same things are broken on trunk.

What do you think @brandonpayton? The only downside is we don't have a flow for preserving in-memory state in other browser tab. This PR would just reload them and destroy their in-memory state. Fixing that would require writing throwaway code for another day, so I suggest merging this destructive strategy and then spending that day on saving temp sites in OPFS.

adamziel commented:
A summary of failed tests:

* Safari: 1 flaky test and 1 legitimate failure when requesting `/wp-admin` instead of `/wp-admin/` – perhaps we need the client-side heuristics to add the trailing slash at least for the wp-admin URL after all. That problem is there in trunk, too.
* Firefox: 2 multisite-related failures. I cannot reproduce them locally. They may or may not be legitimate failures. We don't have a staging environment yet so the only way to find out is deploying and testing that firefox scenario in production. We should get a staging setup really soon.

I'm going to merge this PR with the failing tests without commenting them out. Let them remind us about these suspicious scenarios and let's fix them in the next PR or two.

brandonpayton commented:
> What do you think @brandonpayton? The only downside is we don't have a flow for preserving in-memory state in other browser tab. This PR would just reload them and destroy their in-memory state. Fixing that would require writing throwaway code for another day, so I suggest merging this destructive strategy and then spending that day on saving temp sites in OPFS.

Agreed. I think that's sound reasoning.

brandonpayton commented:
Note: We've disabled the website deployment workflow until temp sites are created in OPFS by default. Then we can be more confident in the Service Worker forcing page reloads when deployments occur.
